From d58a8af45504b59b7404a2d41690f28d2cc4e585 Mon Sep 17 00:00:00 2001
From: Scott Mabin <scott@mabez.dev>
Date: Sat, 12 Sep 2020 23:36:21 +0100
Subject: [PATCH 4/6] Add some Xtensa targets. ESP32, ESP8266 and a generic
 Xtensa target.

Enrich the Xtensa ESP32 and ESP8266 targets with target vendor, env and family

Add some Xtensa targets
---
 compiler/rustc_target/src/spec/mod.rs         |  4 ++
 .../src/spec/xtensa_esp32_none_elf.rs         | 49 +++++++++++++++++++
 .../src/spec/xtensa_esp8266_none_elf.rs       | 49 +++++++++++++++++++
 .../rustc_target/src/spec/xtensa_none_elf.rs  | 47 ++++++++++++++++++
 4 files changed, 149 insertions(+)
 create mode 100644 compiler/rustc_target/src/spec/xtensa_esp32_none_elf.rs
 create mode 100644 compiler/rustc_target/src/spec/xtensa_esp8266_none_elf.rs
 create mode 100644 compiler/rustc_target/src/spec/xtensa_none_elf.rs

diff --git a/compiler/rustc_target/src/spec/mod.rs b/compiler/rustc_target/src/spec/mod.rs
index f949bf95a50..b29b2b278ec 100644
--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -655,6 +655,10 @@ supported_targets! {
 
     ("nvptx64-nvidia-cuda", nvptx64_nvidia_cuda),
 
+    ("xtensa-esp32-none-elf", xtensa_esp32_none_elf),
+    ("xtensa-esp8266-none-elf", xtensa_esp8266_none_elf),
+    ("xtensa-none-elf", xtensa_none_elf),
+
     ("i686-wrs-vxworks", i686_wrs_vxworks),
     ("x86_64-wrs-vxworks", x86_64_wrs_vxworks),
     ("armv7-wrs-vxworks-eabihf", armv7_wrs_vxworks_eabihf),
diff --git a/compiler/rustc_target/src/spec/xtensa_esp32_none_elf.rs b/compiler/rustc_target/src/spec/xtensa_esp32_none_elf.rs
new file mode 100644
index 00000000000..98d2342cdbe
--- /dev/null
+++ b/compiler/rustc_target/src/spec/xtensa_esp32_none_elf.rs
@@ -0,0 +1,49 @@
+use crate::spec::{abi::Abi, LinkerFlavor, PanicStrategy, Target, TargetOptions, RelocModel};
+
+pub fn target() -> Target {
+    Target {
+        llvm_target: "xtensa-none-elf".to_string(),
+        data_layout: "e-m:e-p:32:32-i8:8:32-i16:16:32-i64:64-n32".to_string(),
+        arch: "xtensa".to_string(),
+        pointer_width: 32,
+
+        options: TargetOptions {
+            os_family: Some("unix".to_string()),
+            os: "none".to_string(),
+            env: "newlib".to_string(),
+            vendor: "espressif".to_string(),
+
+            cpu: "esp32".to_string(),
+            endian: "little".to_string(),
+            c_int_width: "32".to_string(),
+            max_atomic_width: Some(32),
+
+            executables: true,
+
+            linker: Some("xtensa-esp32-elf-gcc".to_string()),
+            linker_flavor: LinkerFlavor::Gcc,
+
+            // Because these devices have very little resources having an
+            // unwinder is too onerous so we default to "abort" because the
+            // "unwind" strategy is very rare.
+            panic_strategy: PanicStrategy::Abort,
+
+            // Similarly, one almost always never wants to use relocatable
+            // code because of the extra costs it involves.
+            relocation_model: RelocModel::Static,
+
+            emit_debug_gdb_scripts: false,
+
+            unsupported_abis: vec![
+                Abi::Stdcall,
+                Abi::Fastcall,
+                Abi::Vectorcall,
+                Abi::Thiscall,
+                Abi::Win64,
+                Abi::SysV64,
+            ],
+
+            ..Default::default()
+        },
+    }
+}
diff --git a/compiler/rustc_target/src/spec/xtensa_esp8266_none_elf.rs b/compiler/rustc_target/src/spec/xtensa_esp8266_none_elf.rs
new file mode 100644
index 00000000000..e38f712afc6
--- /dev/null
+++ b/compiler/rustc_target/src/spec/xtensa_esp8266_none_elf.rs
@@ -0,0 +1,49 @@
+use crate::spec::{abi::Abi, LinkerFlavor, PanicStrategy, Target, TargetOptions, RelocModel};
+
+pub fn target() -> Target {
+    Target {
+        llvm_target: "xtensa-none-elf".to_string(),
+        data_layout: "e-m:e-p:32:32-i8:8:32-i16:16:32-i64:64-n32".to_string(),
+        arch: "xtensa".to_string(),
+        pointer_width: 32,
+
+        options: TargetOptions {
+            os_family: Some("unix".to_string()),
+            os: "none".to_string(),
+            env: "newlib".to_string(),
+            vendor: "espressif".to_string(),
+    
+            cpu: "esp8266".to_string(),
+            endian: "little".to_string(),
+            c_int_width: "32".to_string(),
+            max_atomic_width: Some(32),
+
+            executables: true,
+
+            linker: Some("xtensa-lx106-elf-gcc".to_string()),
+            linker_flavor: LinkerFlavor::Gcc,
+
+            // Because these devices have very little resources having an
+            // unwinder is too onerous so we default to "abort" because the
+            // "unwind" strategy is very rare.
+            panic_strategy: PanicStrategy::Abort,
+
+            // Similarly, one almost always never wants to use relocatable
+            // code because of the extra costs it involves.
+            relocation_model: RelocModel::Static,
+
+            emit_debug_gdb_scripts: false,
+
+            unsupported_abis: vec![
+                Abi::Stdcall,
+                Abi::Fastcall,
+                Abi::Vectorcall,
+                Abi::Thiscall,
+                Abi::Win64,
+                Abi::SysV64,
+            ],
+
+            ..Default::default()
+        },
+    }
+}
diff --git a/compiler/rustc_target/src/spec/xtensa_none_elf.rs b/compiler/rustc_target/src/spec/xtensa_none_elf.rs
new file mode 100644
index 00000000000..440c507c4c3
--- /dev/null
+++ b/compiler/rustc_target/src/spec/xtensa_none_elf.rs
@@ -0,0 +1,47 @@
+use crate::spec::{abi::Abi, LinkerFlavor, PanicStrategy, Target, TargetOptions, RelocModel};
+
+pub fn target() -> Target {
+    Target {
+        llvm_target: "xtensa-none-elf".to_string(),
+        data_layout: "e-m:e-p:32:32-i8:8:32-i16:16:32-i64:64-n32".to_string(),
+        arch: "xtensa".to_string(),
+        pointer_width: 32,
+
+        options: TargetOptions {
+            os: "none".to_string(),
+            env: String::new(),
+            vendor: String::new(),
+            
+            endian: "little".to_string(),
+            c_int_width: "32".to_string(),
+            max_atomic_width: Some(32),
+
+            executables: true,
+
+            linker: Some("xtensa-esp32-elf-gcc".to_string()),
+            linker_flavor: LinkerFlavor::Gcc,
+
+            // Because these devices have very little resources having an
+            // unwinder is too onerous so we default to "abort" because the
+            // "unwind" strategy is very rare.
+            panic_strategy: PanicStrategy::Abort,
+
+            // Similarly, one almost always never wants to use relocatable
+            // code because of the extra costs it involves.
+            relocation_model: RelocModel::Static,
+
+            emit_debug_gdb_scripts: false,
+
+            unsupported_abis: vec![
+                Abi::Stdcall,
+                Abi::Fastcall,
+                Abi::Vectorcall,
+                Abi::Thiscall,
+                Abi::Win64,
+                Abi::SysV64,
+            ],
+
+            ..Default::default()
+        },
+    }
+}
\ No newline at end of file
-- 
2.17.1

